singularity-eos/eos/eos_base.hpp

```diff
@@ -148,8 +148,13 @@ class EosBase {
 
   // Generic evaluator
   template <typename Functor_t>
-  constexpr void Evaluate(Functor_t &f) const {
-    CRTP copy = *(static_cast<CRTP const *>(this));
+  PORTABLE_INLINE_FUNCTION void EvaluateDevice(const Functor_t f) const {
+    const CRTP copy = *(static_cast<CRTP const *>(this));
+    f(copy);
+  }
+  template <typename Functor_t>
+  void EvaluateHost(Functor_t &f) const {
+    const CRTP copy = *(static_cast<CRTP const *>(this));
     f(copy);
```

Comment:
We don't really need the explicit copy machinery for the host evaluation, do we? My memory is that this was meant to solve an issue where the object wasn't appropriately copied to device, but I could have misunderstood the issue.

That said, it's possible the compiler returns the same assembly on host regardless of how we write this.

Reply:
I decided to see what the difference was in the canonical CRTP pattern (i.e. calling the base class function which calls the derived implementation) and this was the difference in assembly (GCC 12.3):

copy:
```
 mov    QWORD PTR [rbp-0x8],rdi
 mov    rax,QWORD PTR [rbp-0x8]
 ```
 without an explicit copy:
 ```
 mov    QWORD PTR [rbp-0x18],rdi
 lea    rax,[rbp-0x1]
 ```
 
 From what little I know of assembly, these seem to functionally be the same, right?

Feel free to modify for this exact case, but I'm satisfied that there's no real difference.
https://godbolt.org/z/3vox6on9x

Reply:
We do want the type to be resolved here though to short-circuit inheritance on per-device calls. Will that happen if we don't pull out the type?

Reply:
OOH I see what you mean. Just drop the `copy` var. Yes that's fine.

Reply:
> OOH I see what you mean. Just drop the `copy` var. Yes that's fine.

I mean you tell me which assembly you like better. From my quick test they appear pretty similar to me (maybe `mov` is slightly better than `lea`?) so I'm fine leaving the code as you've written it.

Reply:
Ah I see. I think the assembly's are equivalent. I prefer to leave the code as written just because it fits with the pattern of the rest of the file. But no big deal either way.